src/sgp/Input.cc

```diff
@@ -11,34 +11,79 @@
 
 #include <bitset>
 
+#define BUTTON_REPEAT_TIMEOUT		250
+#define BUTTON_REPEAT_TIME			50
+#define TOUCH_REPEAT_TIMEOUT		750
+#define TOUCH_REPEAT_TIME 			100
 
 // The gfKeyState table is used to track which of the keys is up or down at any one time. This is used while polling
 // the interface.  true = pressed, false = not pressed.
 static std::bitset<2 * SDL_NUM_SCANCODES> gfKeyState;
 static BOOLEAN fCursorWasClipped = FALSE;
 static SGPRect gCursorClipRect;
 
+static void QueuePointerEvent(UINT16 eventType, UINT32 param);
+
+// Bitset to keep track of current mouse buttons and touch finger
+static std::bitset<6> ButtonStates = 0;
+
+// Struct to keep track of a single button or touch state and emit its events
+// The first template parameter indexes into button state
+template<
+	size_t StateIndex,
+	UINT32 BUTTON,
+	UINT16 DOWN_EVENT = MOUSE_BUTTON_DOWN,
+	UINT16 UP_EVENT = MOUSE_BUTTON_UP,
+	UINT16 REPEAT_EVENT = MOUSE_BUTTON_REPEAT,
+	UINT32 REPEAT_TIMEOUT = BUTTON_REPEAT_TIMEOUT,
+	UINT32 REPEAT_TIME = BUTTON_REPEAT_TIME
+>
+struct ButtonState {
+	private:
+		UINT32 repeatTimer;
+
+	public:
+		void handleDown() {
+			repeatTimer = GetClock() + REPEAT_TIMEOUT;
+			ButtonStates[StateIndex] = TRUE;
+			QueuePointerEvent(DOWN_EVENT, BUTTON);
+		}
+
+		void handleUp() {
+			repeatTimer = 0;
+			ButtonStates[StateIndex] = FALSE;
+			QueuePointerEvent(UP_EVENT, BUTTON);
+		}
+
+		void detectRepeat(UINT32 uiTimer) {
+			if (isDown()) {
+				if (repeatTimer > 0 && repeatTimer <= uiTimer)
+				{
+					QueuePointerEvent(REPEAT_EVENT, BUTTON);
+					repeatTimer = uiTimer + REPEAT_TIME;
+				}
+			} else {
+				repeatTimer = 0;
+			}
+		}
 
-// These data structures are used to track the mouse while polling
-static UINT32 guiLeftButtonRepeatTimer;
-static UINT32 guiRightButtonRepeatTimer;
-static UINT32 guiMiddleButtonRepeatTimer;
-static UINT32 guiX1ButtonRepeatTimer;
-static UINT32 guiX2ButtonRepeatTimer;
-static UINT32 guiFingerRepeatTimer;
+		BOOLEAN isDown() const {
+			return ButtonStates[StateIndex];
+		}
+};
 
-BOOLEAN gfLeftButtonState;  // TRUE = Pressed, FALSE = Not Pressed
-BOOLEAN gfRightButtonState; // TRUE = Pressed, FALSE = Not Pressed
-BOOLEAN gfMiddleButtonState;// TRUE = Pressed, FALSE = Not Pressed
-BOOLEAN gfX1ButtonState;// TRUE = Pressed, FALSE = Not Pressed
-BOOLEAN gfX2ButtonState;// TRUE = Pressed, FALSE = Not Pressed
+static ButtonState<0, MOUSE_BUTTON_LEFT> gLeftButtonState;
+static ButtonState<1, MOUSE_BUTTON_RIGHT> gRightButtonState;
+static ButtonState<2, MOUSE_BUTTON_MIDDLE> gMiddleButtonState;
+static ButtonState<3, MOUSE_BUTTON_X1> gX1ButtonState;
+static ButtonState<4, MOUSE_BUTTON_X2> gX2ButtonState;
+static ButtonState<5, 0, TOUCH_FINGER_DOWN, TOUCH_FINGER_UP, TOUCH_FINGER_REPEAT, TOUCH_REPEAT_TIMEOUT, TOUCH_REPEAT_TIME> gMainFingerState;
```

Comment:
The mouse ones could still be grouped into an array or somesuch, indexed by button type. Then IsMouseButtonDown wouldn't need an extra mapping.
Ideally it would be in one object, so we could use the bitset capabilities, eg. merging `IsMouseButtonDown(MOUSE_BUTTON_LEFT) || (IsMouseButtonDown(MOUSE_BUTTON_RIGHT)` to `IsMouseButtonDown(MOUSE_BUTTON_LEFT | MOUSE_BUTTON_RIGHT)`.

Reply:
While I agree that using bitflags for the individual mouse buttons would be nicer from the user perspective, I would not optimize this further right now. I didn't find any occurence of checking multiple mouse buttons in a single condition in our current codebase (that didn't also check for other things).

From my current understanding of the input/mouse system, using `IsMouseButtonDown` indicates code smell anyways (didn't want to setup a mouse region or didn't use the information of the mouse region). Same for usage of `gusMouseXPos`, `gusMouseYPos`. All of this information is propagated through the mouse regions already, it just needs to be used from there:

- `ButtonState` contains mouse button state using bitflags (so there we can already check multiple buttons)
- `MouseXPos`, `MouseYPos` contains absolute mouse coordinates

Reply:
That (shortened) example is from WaitForHelpScreenResponse. But yeah, it doesn't have to be done within the same PR, another issue could be opened as a reminder that we have needlessly competing implementations and only one should stay.

Reply:
I give you this one :grinning:. THB I excluded searching through the editor files, there is a bunch of weird input handling in there. I have opened https://github.com/ja2-stracciatella/ja2-stracciatella/issues/1557. I will also include other things that I notice along the way...